clients/csharp-uwp/UWPVoiceAssistantSample/DirectLineSpeechDialogBackend.cs

```diff
@@ -395,7 +341,163 @@ private DialogServiceConfig CreateConfiguration()
             return config;
         }
 
-        private bool TryRefreshConfigValues()
+        private void InitializeKeywordRecognizer()
+        {
+            this.keywordAudioConfig = AudioConfig.FromStreamInput(this.audioIntoKeywordSink);
+
+            // Disable throttling of input audio (send it as fast as we can!)
+            this.keywordAudioConfig.SetProperty("SPEECH-AudioThrottleAsPercentageOfRealTime", "9999");
+            this.keywordAudioConfig.SetProperty("SPEECH-TransmitLengthBeforThrottleMs", "10000");
+
+            if (LocalSettingsHelper.SpeechSDKLogEnabled)
+            {
+                var logPath = $"{ApplicationData.Current.LocalFolder.Path}\\SpeechSDK.log";
+                this.keywordAudioConfig.SetProperty(PropertyId.Speech_LogFilename, logPath);
+            }
+
+            this.keywordRecognizer = new KeywordRecognizer(this.keywordAudioConfig);
+
+            this.keywordRecognizer.Recognized += (_, recoArgs) =>
+            {
+                this.logger.Log($"Keyword confirmed on device (@{this.audioIntoKeywordSink.AudioReadSinceReset.TotalMilliseconds:0}ms): {recoArgs.Result.Text}");
+                this.audioIntoKeywordSink.BookmarkPosition = TimeSpan.Zero;
+                this.KeywordRecognizing?.Invoke(recoArgs.Result.Text);
+                this.audioIntoConnectorSink.DataSource = PullAudioDataSource.FromKeywordResult(recoArgs.Result);
+                this.audioIntoConnectorSink.BookmarkPosition = KeywordRejectionTimeout;
+                this.EnsureConnectorReady();
+                this.logger.Log($"Starting connector");
+                _ = this.connector.StartKeywordRecognitionAsync(this.ConfirmationModel as KeywordRecognitionModel);

```

Comment:
@trrwilson  Does this line mean 2nd stage recognition will be run a second time? If so, is there a way to tell the connector that 2nd stage has already been evaluated and to skip to 3rd stage? 

(I presume if 3rd stage isn't enabled, then a call to ListenOnceAsync() would work here, but that StartKeywordRecognitionAsync is being used instead to ensure 3rd stage gets called)

Reply:
Hey @tomh05! Thanks for piling on here and apologies I've been away a bit--between a week of vacation and then getting pulled into a few things, I've been more absent on this front than I'd like.

You are 100% correct that the DialogServiceConnector will unnecessarily repeat the on-device portion of the keyword spotting. Very rarely (but still occasionally) they'll even disagree, with the DialogServiceConnector not deigning to fire after the KeywordRecognizer does. That last part is likely attributable to subtle differences in the byte alignment based on how the KeywordRecognitionResult selects the audio start position, but I'm rambling.

Two parts of that:
* The SDK should support a means of doing this. We have an item on our backlog to design and implement a way to chain a KeywordRecognizer and DialogServiceConnector together to get KWV without repeating on-device KWS. I'd be curious to know from you as one of our most informed consumers, though: how do you think it should work, i.e. what code would make sense to write? There are a lot of ways to specify this--something in the config, a parameter to the Start() call, a different way of creating/annotating the input, and more--and part of the debate is what the most intuitive and clear way to expose this option/capability would be.
* In the interim, one thing that can be experimented with is using ListenOnce instead of KWS on the DialogServiceConnector. That *will not* be eligible for KWV right now, but at least as an exploration/prototyping step, it'd allow a full observation of what latency benefits skipping that second step would have. In my own ad hoc testing for this change, I saw that the KWS delayed the start of stream to the service by ~200-400ms depending on configuration; that doesn't translate to a full 200-400ms of extra time until the KWV result arrives (it runs faster than real-time by a good margin), but it's still going to be a considerable increase.

Reply:
Hi @trrwilson! Will throw my 2 cents in here 🙂 

Re: SDK provision for avoiding the repition of 2nd stage, this feels like a good fit for a setting on the Connector, because it should be consistent in an application based on the architecture. Adding a parameter on the `Start` call would imply that it could be different at the initialization of different conversations within the same app, but surely any given app would either use the `KeywordRecognizer`, or it wouldn't? If there is a usecase for apps to switch between online and offline recognition then a parameter to the `Start` call seems like a reasonable fallback position 👍 

Since we can't do 3rd stage verification just now we can experiment with using `ListenOnce` in the short term and see what the latency improvements are like. Is there a timeline for the SDK changes to avoid repeating 2nd stage? Once 3rd stage is available to us it would be a shame to have to pick between "have 3rd stage, but gain latency from repeating 2nd" and "lower latency, but no 3rd stage".